### ISSUE TITLE
Awaiting readback completion without spinning

### ISSUE THREAD

**Issue by endragor** [OurMachinery/themachinery-public#919]
When computing things on the GPU in a separate thread (briefly described in #918) you don't want to burn CPU by spinning in a read_complete() loop because that would wastefully consume resources of the frame loop. Instead, you want to "sleep until GPU tasks are done". [vkWaitForFences](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkWaitForFences.html) allows to do exactly this, but it is not exposed in the renderer API.

**Comment by frank-ourmachinery** [comment]
The readback system currently in place has to keep track of all in-flight reads. This needs to be thread-safe in order for multiple fibres to check for readback completion which in turn requires a mutex. If we were to allow users to wait for fences directly we would either block all other threads from reading during that period.
There are other solutions to this, but as far as I can see this is more of a design decision. We prefer not to wait on anything if at all possible and instead yield the active thread to do some other work.

**Comment by endragor** [comment]
It's been a while since I last used The Machinery so my memory is a bit rusty here. But how does async computation on the GPU fit in the current design? "Yielding the active thread" is a concept related to the job system and the job system is tied to the frame loop. But there are things that you want to do outside of the frame loop, such as AI or preparing distant parts of an open world as the player approaches them. I had problems using The Machinery with these use cases and filed a bunch of related issues here.

**Comment by niklas-ourmachinery** [comment]
For things like that (AI or preparing distant parts of the world)... I would probably try to chop it up into "frame sized" chunks and execute one of them every frame.

We also have the "task system" which is intended for long-running tasks and executes on a background thread. It can be used for things like background reads from disk, etc. You *could* use that for your examples too, but it's not ideal for tasks like this that are purely computational. If you do, you basically say to the OS: "I don't know how much of this work that should be scheduled each frame -- you figure it out!"

Modern OSs tend to be written with interactive desktop applications in mind, not games, so they don't care about you hitting 60 Hz. So the OS may decide to give the "background AI task" 10 ms of CPU time and cause your game to hitch a frame. You can try to fix it with priorities, but then you can run into the other problem where your AI tasks never complete on time.

In conclusion, I generally think it's preferrable for *you* to take control of how much work should be done on each task each frame by chopping them up in suitably sized pieces.

**Comment by endragor** [comment]
I agree about such approach for CPU-based compute work, but this issue is about computations on the GPU which from the CPU side are conceptually similar to waiting for background reads from disk. I didn't find a way to submit (asynchronous) tasks to the GPU and then await them without consuming unreasonable amounts of CPU resources with The Machinery.
